lgc/builder/BuilderImpl.cpp

```diff
@@ -74,6 +74,165 @@ Value *BuilderImplBase::CreateDotProduct(Value *const vector1, Value *const vect
+//
+// @param vector1 : The integer Vector 1
+// @param vector2 : The integer Vector 2
+// @param accumulator : The accumulator to the scalar of dot product
+// @param flags : The first bit marks whether Vector 1 is signed and the second bit marks whether Vector 2 is signed
+// @param instName : Name to give instruction(s)
+Value *BuilderImplBase::CreateIntegerDotProductNew(Value *vector1, Value *vector2, Value *accumulator, unsigned flags,
+                                                   const Twine &instName) {
+  if (flags == SecondVectorSigned) {
+    std::swap(vector1, vector2);
+    flags = FirstVectorSigned;
+  }
+  const bool isBothSigned = (flags == (FirstVectorSigned | SecondVectorSigned));
+  const bool isMixedSigned = (flags == FirstVectorSigned);
+  const bool isSigned = isBothSigned || isMixedSigned;
+
+  // The factor inputs are always <N x iM> of the same type
+  Type *inputTy = vector1->getType();
+  assert(inputTy->isVectorTy() && inputTy->getScalarType()->isIntegerTy() && inputTy == vector2->getType());
+  const unsigned compCount = cast<FixedVectorType>(inputTy)->getNumElements();
+
+  // The supported size of M
+  const unsigned compBitWidth = inputTy->getScalarSizeInBits();
+  assert(compBitWidth == 4 || compBitWidth == 8 || compBitWidth == 16 || compBitWidth == 32 || compBitWidth == 64);
+
+  // The result type is given by accumulator, which must be greater than or equal to that of the components of Vector 1
+  Type *resultTy = accumulator->getType();
+  const bool hasAccumulator = !(isa<ConstantInt>(accumulator) && cast<ConstantInt>(accumulator)->isNullValue());
+  const unsigned resBitWidth = resultTy->getScalarSizeInBits();
+  assert(resBitWidth == 4 || resBitWidth == 8 || resBitWidth == 16 || resBitWidth == 32 || resBitWidth == 64);
+
+  // Check if there is a native intrinsic that can do the entire operation (dot product and saturating accumulate) in a
+  // single instruction. They must meet the two conditions:
+  // 1. The required native intrinsic is supported by the specified hardware
+  // 2. The factor inputs must be <2 x i16> or <N x i8> (N <= 4) or <N x i4> (N <= 8)
+  const auto &supportIntegerDotFlag = getPipelineState()->getTargetInfo().getGpuProperty().supportIntegerDotFlag;
+  const bool isSupportCompBitwidth = (supportIntegerDotFlag.compBitwidth16 && compBitWidth == 16) ||
+                                     (supportIntegerDotFlag.compBitwidth8 && compBitWidth == 8) ||
+                                     (supportIntegerDotFlag.compBitwidth4 && compBitWidth == 4);
+  const bool isSupportSignedness =
+      isMixedSigned ? supportIntegerDotFlag.diffSignedness : supportIntegerDotFlag.sameSignedness;
+  const bool isDot2 = (compCount == 2 && compBitWidth == 16);
+  const bool isDot4 = (compCount <= 4 && compBitWidth == 8);
+  const bool isDot8 = (compCount <= 8 && compBitWidth == 4);
+  const bool hasNativeIntrinsic = isSupportCompBitwidth && isSupportSignedness && (isDot2 || isDot4 || isDot8);
```

Comment:
This also needs to check that `resultBitWidth <= 32`.

Reply:
> However, to be clear, the final version of the change should replace CreateIntegerDotProduct entirely with the new code. Having both CreateIntegerDotProduct and CreateIntegerDotProductNew would just be cruft that adds more code for no good reason.

Sure, I'm going to make a new PR to switch to the new code path after the internal change for mixed cases is submitted.

Reply:
Done.